### PR TITLE
CORDA-1997 Added constraint type information to vault states table.

### PR DESCRIPTION
**Pull request by josecoll**
Vault storage of contract state constraints metadata and associated vault query functions to retrieve and sort by constraint type.

Note: a follow-up PR will address performance improvements as per
https://r3-cev.atlassian.net/browse/CORDA-2013

### CODE REVIEW

File: node/src/main/resources/migration/vault-schema.changelog-v6.xml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.1" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.5.xsd">
+    <changeSet author="R3.Corda" id="add_is_constraint_information_columns">
+        <preConditions onFail="MARK_RAN"><not><columnExists tableName="vault_states" columnName="is_relevant"/></not></preConditions>
```
Comment (tudor-malene) [inline node/src/main/resources/migration/vault-schema.changelog-v6.xml]: what's up with this precondition?

File: node/src/main/resources/migration/vault-schema.changelog-v6.xml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+        <addColumn tableName="vault_states">
+            <column name="constraint_type" type="INT"/>
+	    </addColumn>
+	    <update tableName="vault_states">
+	        <column name="constraint_type" valueNumeric="0"/>
```
Comment (tudor-malene) [inline node/src/main/resources/migration/vault-schema.changelog-v6.xml]: You can write this more compact.
see:
https://stackoverflow.com/questions/11042247/when-creating-a-column-using-liquibase-how-do-i-specify-a-value-for-that-column

File: core/src/main/kotlin/net/corda/core/node/services/VaultService.kt [hunk 3]
```diff
@@ -125,6 +127,45 @@ class Vault<out T : ContractState>(val states: Iterable<StateAndRef<T>>) {
...
+    @CordaSerializable
+    data class ConstraintInfo(val constraint: AttachmentConstraint) {
+        @CordaSerializable
+        enum class Type {
+            ALWAYS_ACCEPT, HASH, CZ_WHITELISTED, SIGNATURE
+        }
+        fun type(): Type {
+            return when (constraint::class.java) {
+                AlwaysAcceptAttachmentConstraint::class.java -> Type.ALWAYS_ACCEPT
+                AutomaticHashConstraint::class.java,
```
[the remark this answers is not in the document]
Comment (josecoll) [reply core/src/main/kotlin/net/corda/core/node/services/VaultService.kt]: Sure - I can remove it, but code processing around `AutomaticHashConstraint` currently seems an unnecessary indirection to `HashConstraint` anyway.
eg. shouldn't there be some `AutomaticConstraint` logic that just defaults to the preferred constraint? (which should be `SignatureConstraint` for v4.x)

File: node/src/main/resources/migration/vault-schema.changelog-v6.xml (hunk 1 above)
Comment (josecoll) [reply node/src/main/resources/migration/vault-schema.changelog-v6.xml]: removed

**Comment by josecoll** [comment]
I will extend the query to query by constraint data also.

File: core/src/main/kotlin/net/corda/core/node/services/VaultService.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (tudor-malene) [reply core/src/main/kotlin/net/corda/core/node/services/VaultService.kt]: Exactly that was done on a different PR: It's called AutomaticPlaceholderConstraint.
But neither this new one or the deprecated AutomaticHashConstraint shouldn't be in the vault.

File: core/src/main/kotlin/net/corda/core/node/services/VaultService.kt [hunk 4]
```diff
@@ -184,6 +225,30 @@ class Vault<out T : ContractState>(val states: Iterable<StateAndRef<T>>) {
         ): StateMetadata {
             return StateMetadata(ref, contractStateClassName, recordedTime, consumedTime, status, notary, lockId, lockUpdateTime, null)
         }
+        constructor(ref: StateRef,
```
Comment (tudor-malene) [inline core/src/main/kotlin/net/corda/core/node/services/VaultService.kt]: Could this be done with a default value and JvmOverloads?
I guess the same can be said about the relevancy

File: core/src/main/kotlin/net/corda/core/node/services/VaultService.kt [hunk 5]
```diff
@@ -160,18 +200,9 @@ class Vault<out T : ContractState>(val states: Iterable<StateAndRef<T>>) {
             val notary: AbstractParty?,
             val lockId: String?,
             val lockUpdateTime: Instant?,
-            val relevancyStatus: Vault.RelevancyStatus?
+            val relevancyStatus: Vault.RelevancyStatus? = null,
+            val constraintInfo: ConstraintInfo = ConstraintInfo(AlwaysAcceptAttachmentConstraint)
```
Comment (tudor-malene) [inline core/src/main/kotlin/net/corda/core/node/services/VaultService.kt]: This needs to be nullable.
see: https://docs.corda.net/serialization-default-evolution.html#adding-non-nullable-properties

**Comment by josecoll** [comment]
@tudor-malene back to you ...

**Pull request merged**